swap/swap.go

```diff
@@ -434,86 +436,86 @@ func (s *Swap) Balances() (map[enode.ID]int64, error) {
-	return cheque, err
-}
-
-// SentCheques returns the last sent cheques for all known SWAP peers
-func (s *Swap) SentCheques() (map[enode.ID]*Cheque, error) {
-	cheques := make(map[enode.ID]*Cheque)
+// Cheques returns all known last sent and received cheques, grouped by peer
+func (s *Swap) Cheques() (map[enode.ID]map[string]*Cheque, error) {
+	cheques := make(map[enode.ID]map[string]*Cheque)
 
+	// get peer cheques from memory
 	s.peersLock.Lock()
 	for peer, swapPeer := range s.peers {
 		swapPeer.lock.Lock()
-		cheques[peer] = swapPeer.getLastSentCheque()
+		cheques[peer] = make(map[string]*Cheque)
+		cheques[peer][sentChequeResponseKey] = swapPeer.getLastSentCheque()
+		cheques[peer][receivedChequeResponseKey] = swapPeer.getLastReceivedCheque()
 		swapPeer.lock.Unlock()
 	}
 	s.peersLock.Unlock()
 
-	// add store cheques, if peer was not already added
-	chequesIterFunction := func(key []byte, value []byte) (stop bool, err error) {
-		peer := keyToID(string(key), sentChequePrefix)
-		if _, peerHasCheque := cheques[peer]; !peerHasCheque {
-			var peerCheque Cheque
-			err = json.Unmarshal(value, &peerCheque)
-			if err == nil {
-				cheques[peer] = &peerCheque
-			}
-		}
-		return stop, err
+	// get peer cheques from store
+	err := s.addStoreCheques(sentChequePrefix, sentChequeResponseKey, cheques)
+	if err != nil {
+		return nil, err
 	}
-	err := s.store.Iterate(sentChequePrefix, chequesIterFunction)
+	err = s.addStoreCheques(receivedChequePrefix, receivedChequeResponseKey, cheques)
 	if err != nil {
 		return nil, err
 	}
 
-	return cheques, nil
-}
-
-// ReceivedCheque returns the last received cheque for a given peer
-func (s *Swap) ReceivedCheque(peer enode.ID) (cheque *Cheque, err error) {
-	if swapPeer := s.getPeer(peer); swapPeer != nil {
-		return swapPeer.getLastReceivedCheque(), nil
+	// fill in result with missing cheques
+	for _, peerCheques := range cheques {
+		// add nil as type of cheque if not present
+		if _, peerHasReceivedCheque := peerCheques[receivedChequeResponseKey]; !peerHasReceivedCheque {
+			peerCheques[receivedChequeResponseKey] = nil
+		}
+		if _, peerHasSentCheque := peerCheques[sentChequeResponseKey]; !peerHasSentCheque {
+			peerCheques[sentChequeResponseKey] = nil
+		}
 	}
-	err = s.store.Get(receivedChequeKey(peer), &cheque)
-	return cheque, err
-}
-
-// ReceivedCheques returns the last received cheques for all known SWAP peers
-func (s *Swap) ReceivedCheques() (map[enode.ID]*Cheque, error) {
-	cheques := make(map[enode.ID]*Cheque)
 
-	s.peersLock.Lock()
-	for peer, swapPeer := range s.peers {
-		swapPeer.lock.Lock()
-		cheques[peer] = swapPeer.getLastReceivedCheque()
-		swapPeer.lock.Unlock()
-	}
-	s.peersLock.Unlock()
+	return cheques, nil
+}
 
-	// add store cheques, if peer was not already added
+// add disk cheques for peers not already present in given cheques map
+func (s *Swap) addStoreCheques(chequePrefix string, chequeKey string, cheques map[enode.ID]map[string]*Cheque) error {
```

Comment:
`(chequePrefix, chequeKey string .....)`
also, i'm not sure why do you need to name parameters in this way - is it not easier to just build these prefixes/keys inside the function according to the underlying type which is passed? (the same way that you construct the `peer` key below)

Reply:
in swap we deal with sent and received cheques in a similar manner: both are put in the store, but with different prefixes (remember the discussion about iterating through the store?)

in order to minimize duplicated code, i tried to build a function that would be useful for both types of cheques.

here is the call in question:
```golang
// get peer cheques from store
err := s.addStoreCheques(sentChequePrefix, sentChequeResponseKey, cheques)
if err != nil {
	return nil, err
}
err = s.addStoreCheques(receivedChequePrefix, receivedChequeResponseKey, cheques)
if err != nil {
	return nil, err
}
```
- the `chequePrefix` is used to iterate through the store for a specific type of cheque.
- the `chequeKey` is used to place found cheques inside of the `cheques` map.

now, going back to your original comment: how does "the underlying type" look like, if it were passed to this function? is it a string? 

and how would we construct the prefixes and keys? a `map`, a `switch`?